### DISCUSSION TITLE
How can I diagnose/improve very slow test discovery?

### DISCUSSION THREAD

**Discussion by alx9r** [xunit/xunit#2165] [Question] [answered]
I have two projects

- _impl_
   * does not reference xunit
   * contains a fairly elaborate hierarchy of generic types
- _test_
   * contains approximately 30 tests
   * references packages xunit and xunit.runner.visualstudio
   * references project _impl_

During a recent refactoring of _impl_ the test discovery time increased from 8s to 17s.  There were discovery times that were far longer along the way; I tagged a commit where a related project had a test discovery time of over 5 minutes.  I'd like to better understand what is happening during discovery so that I can hopefully avert the long discovery times.

_test_ didn't change much during the refactoring.  Many types in _impl_ had generic parameters added during the refactoring.

I'm using the following command to measure test discovery time:

    dotnet clean && dotnet test && Measure-Command { dotnet test -t }

I've experimented a bit with `dotnet test --filter` but that doesn't seem to affect discovery.

1. What is xunit doing during test discovery here?  
2. What could be happening that for one refactoring of a project under test discovery takes minutes, while for a similar refactoring takes only seconds?
3. Can a narrower discovery search scope be specified?

**Comment by alx9r** [comment, marked as answer]
I compared `dotnet test -t` execution times with nunit for the tag where xunit test discovery took ~5minutes.  I used NUnit.Framework instead of XUnit and transformed the tests to idiomatic nunit.  The results were as follows:

| Framework | discovery time |
|---|---|
| xunit | 314s |
| nunit | 6.7s |

**Reply by bradwilson** [reply]
Can you provide the repro projects? /cc @clairernovotny

**Reply by alx9r** [reply]
@bradwilson Currently I cannot.  But I'm hoping to find a way to extract a repro project that I can share.  My crude attempts to do so have resulted in the problem going away.  I'll have to take a more surgical approach I think.  I was hoping to have a better understanding of what happens with the test assemblies during `dotnet test -t` before I embark on that.

**Reply by bradwilson** [reply]
Going to mark this as answered since there is no repro project for us to work from. If you can find a repro project, please open an issue (instead of a discussion) and I'll take a look at it. Thanks!

**Reply by jimiscott** [reply]
@bradwilson We too are experiencing very slow discovery. On a large solution of about 100 projects, trying to run a single test in VS2019 typically takes about 35 seconds to complete discovery. 

Are there any diagnostics which would be helpful to understand the cause of the issue?

**Reply by bradwilson** [reply]
@jimiscott Without a repro project I'm not really sure what's going on.

One thing you can try is [disabling theory pre-enumeration](https://xunit.net/docs/configuration-files#preEnumerateTheories) to see if the issue is related to large/expensive amounts of theory data.

**Reply by jimiscott** [reply]
@bradwilson Thanks for your reply. We have (finally) found the solution. We had to exclude the following two processes from Windows Defender/Antivirus:

- testhost.net48.x86.exe (this would need to be changed dependent on framework version and bitness respectively) AND;
- ServiceHub.RoslynCodeAnalysis.exe 

(It is slightly strange that our MSTest tests do not exhibit the pronounced lag that xUnit tests do).

**Reply by bradwilson** [reply]
Interesting to note. One of the first thing I do with a fresh box is exclude my source folder(s) from Defender, which is why I've never noted this. As for why this affects us and not MSTest... your guess is as good as mine. The architectures are entirely different, so it would require some extensive investigation to discover what the actual performance interaction is.

**Reply by jimiscott** [reply]
Brad, our various solution directories are excluded.

This perf issue only seems to be affecting our core solution (100+ projects) with approx. 5500+ tests (MSTest & Xunit).

I will leave it there, hoping this helps others.

**Reply by Maximys** [reply]
@bradwilson , I have the same problem with https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime.Numerics project. Can you profile xunit with it?
